File: flink_java/src/main/java/github/yahuili1128/streamOperators/SelectDemo.java
package github.yahuili1128.streamOperators;

import github.yahuili1128.connector.SourceKafka010;
import github.yahuili1128.pojo.MockUpModel;
import org.apache.flink.api.java.typeutils.runtime.PojoSerializerSnapshot;
import org.apache.flink.streaming.api.collector.selector.OutputSelector;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.datastream.SplitStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

import java.util.ArrayList;

/**
 * @Description :  flink select java
 * @Author : LiYahui
 * @Date : 2019-07-15 13:36
 * @Version : V1.0
 */
public class SelectDemo {
	public static void main(String[] args) throws Exception {
		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
		SingleOutputStreamOperator<MockUpModel> mockUpkafka010 = SourceKafka010.getMockUpkafka010(env);
		SplitStream<MockUpModel> splitDS = mockUpkafka010.split(new OutputSelector<MockUpModel>() {
			@Override
			public Iterable<String> select(MockUpModel value) {
				ArrayList<String> output = new ArrayList<>();
				if (value.gender.equals("male")) {
					output.add("male");
				} else {
					output.add("famle");
				}

				return output;
			}
		});
		//		splitDS.print().setParallelism(1);
		//		MockUpModel(name=yahui, gender=female, timestamp=1563168888165, age=29)
		// select 算子已经过期
		DataStream<MockUpModel> maleDS = splitDS.select("male");
		DataStream<MockUpModel> famleDS = splitDS.select("famle");
		maleDS.print().setParallelism(1);
		famleDS.print().setParallelism(1);

		env.execute("kafka010 demo");
	}
}

File: flink_scala/src/main/scala/github/yahuili1128/streamOperators/SelectDemo.scala
package github.yahuili1128.streamOperators

import github.yahuili1128.connector.ScalaSourceKafka010
import github.yahuili1128.connector.ScalaSourceKafka010.MockUp
import org.apache.flink.streaming.api.scala.{DataStream, SplitStream, StreamExecutionEnvironment}

/**
  * @Description : flink select scala
  * @Author : LiYahui
  * @Date : 2019-07-15 18:35
  * @Version : V1.0
  */
object SelectDemo {
	def main(args: Array[String]): Unit = {
		val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
		val mockup: DataStream[ScalaSourceKafka010.MockUp] = ScalaSourceKafka010.getMockUpFromKafka(env)
		val splitDS: SplitStream[MockUp] = mockup.split((line: MockUp) => line.gender match {
			case "male" => List("male")
			case "female" => List("female")
		})
		splitDS.select("male")
				.print()
				.setParallelism(1)

		env.execute(s"${this.getClass.getSimpleName}")
	}

}
